### PR TITLE
[ALLUXIO-2232] Reworking File System REST API.

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses: https://alluxio.atlassian.net/browse/ALLUXIO-2232

This PR changes the REST API to better follow the RESTful principles (endpoints are resources not actions).

This PR decouples the metadata operations of file create / open / close from the data operation of reading / writing a file. To do so, it introduces adds stream resources to the API and implements a stream cache in the proxy.

### CODE REVIEW

**Comment by jsimsa** [comment]
Here is visualization of the new API surface and example of endpoint documentation:
screen shot 2016-11-30 at 11 17 39 am

File: core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java [hunk 1]
```diff
@@ -113,11 +125,19 @@ public Mode getMode() {
   /**
    * @return the under storage type
    */
+  @JsonIgnore
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java]: If `WriteType` supports `getUnderStorageType` and we are exposing `WriteType`, we do not need to have separate getters for the Alluxio/UnderStorage type

**Comment by jsimsa** [comment]
\o/

File: core/common/src/main/java/alluxio/Constants.java [hunk 2]
```diff
@@ -111,9 +110,8 @@
   public static final String FILE_SYSTEM_WORKER_CLIENT_SERVICE_NAME = "FileSystemWorkerClient";
   public static final String KEY_VALUE_MASTER_CLIENT_SERVICE_NAME = "KeyValueMasterClient";
   public static final String KEY_VALUE_WORKER_CLIENT_SERVICE_NAME = "KeyValueWorkerClient";
-  public static final String FILE_SYSTEM_CLIENT_SERVICE_NAME = "FileSystemClient";
 
-  public static final String REST_API_PREFIX = "/v1/api";
+  public static final String REST_API_PREFIX = "/api/v1";
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/Constants.java]: Why reverse the two?

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 3]
```diff
@@ -89,407 +98,375 @@ public FileSystemClientRestServiceHandler(@Context ServletContext context) {
...
-   * @summary get the service version
-   * @return the response object
-   */
-  @GET
-  @Path(SERVICE_VERSION)
-  @ReturnType("java.lang.Long")
-  public Response getServiceVersion() {
-    return RestUtils.call(new RestUtils.RestCallable<Long>() {
+  @POST
+  @Path(STREAMS_PREFIX + ID_PARAM + CLOSE)
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: this is sort of fragile since it assumes that each of these strings have compatible "/" startings/endings. For example, we couldn't use both path and id params with `STREAMS_PREFIX + PATH_PARAM + ID_PARAM + CLOSE`. I think it would be less fragile if we just wrote out the string, e.g. "streams/{id}/close".

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 4]
```diff
@@ -89,407 +98,375 @@ public FileSystemClientRestServiceHandler(@Context ServletContext context) {
...
+  @ReturnType("java.lang.Void")
+  public Response close(@PathParam("id") final Integer id) {
+    return RestUtils.call(new RestUtils.RestCallable<Void>() {
       @Override
-      public Long call() throws Exception {
-        return Constants.FILE_SYSTEM_CLIENT_SERVICE_VERSION;
+      public Void call() throws Exception {
+        FileInStream is = sInStreams.get(id);
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: could simplify this a little by using `remove` instead of `get`

File: core/common/src/main/java/alluxio/Constants.java (hunk 2 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Constants.java]: as per suggestion by @aaudiber (I am fine with either and not reversing means note braking backwards compatibility with 1.3 REST API clients)

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 5]
```diff
@@ -89,407 +98,375 @@ public FileSystemClientRestServiceHandler(@Context ServletContext context) {
...
+  public Response createDirectory(@PathParam("path") final String path,
+      final CreateDirectoryOptions options) {
     return RestUtils.call(new RestUtils.RestCallable<Void>() {
       @Override
       public Void call() throws Exception {
-        Preconditions.checkNotNull(path, "required 'path' parameter is missing");
-        CreateDirectoryOptions options = CreateDirectoryOptions.defaults();
-        if (allowExists != null) {
-          options.setAllowExists(allowExists);
+        if (options == null) {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: (optional) could use the pattern
```java
if (options == null) {
  options = CreateDirectoryOptions.defaults();
}
mFileSystem.createDirectory(new AlluxioURI(path), options);
```

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 6]
```diff
@@ -89,407 +98,375 @@ public FileSystemClientRestServiceHandler(@Context ServletContext context) {
...
-        mFileSystem.delete(new AlluxioURI(path), options);
-        return null;
+      public Integer call() throws Exception {
+        FileOutStream os;
+        if (options == null) {
+          os = mFileSystem.createFile(new AlluxioURI(path), CreateFileOptions.defaults());
+        } else {
+          os = mFileSystem.createFile(new AlluxioURI(path), options);
+        }
+        sOutStreams.put(os.hashCode(), os);
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: maybe increment a synchronized counter to avoid hash collisions?

File: core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java (hunk 1 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java]: done

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 7]
```diff
@@ -89,407 +98,375 @@ public FileSystemClientRestServiceHandler(@Context ServletContext context) {
...
-  public Response upload(@QueryParam("path") final String path,
-      @QueryParam("blockSizeBytes") final Long blockSizeBytes,
-      @QueryParam("locationPolicy") final String locationPolicy,
-      @QueryParam("mode") final Integer mode,
-      @QueryParam("recursive") final Boolean recursive,
-      @QueryParam("ttl") final Long ttl,
-      @QueryParam("ttlAction") final TtlAction ttlAction,
-      @QueryParam("writeType") final WriteType writeType,
+  public Response write(@PathParam("id") final Integer id,
       final InputStream is) {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: does this need to wrap?

File: tests/src/test/java/alluxio/rest/TestCase.java [hunk 8]
```diff
@@ -127,11 +128,13 @@ public String call() throws Exception {
       connection.setRequestProperty("Content-Type", "application/octet-stream");
       ByteStreams.copy(mOptions.getInputStream(), connection.getOutputStream());
     }
-    if (mOptions.getJsonString() != null) {
+    if (mOptions.getBody() != null) {
       connection.setDoOutput(true);
-      connection.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
+      connection.setRequestProperty("Content-Type", "application/json");
+      ObjectMapper mapper = new ObjectMapper();
+      mapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/rest/TestCase.java]: why is this needed?

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 9]
```diff
@@ -59,24 +58,34 @@
...
   public static final String RENAME = "rename";
-  public static final String SET_ATTRIBUTE = "set_attribute";
+  public static final String SET_ATTRIBUTE = "set-attribute";
   public static final String UNMOUNT = "unmount";
-  public static final String UPLOAD = "upload";
+  public static final String WRITE = "write";
 
   private final FileSystem mFileSystem;
+  private static Map<Integer, FileOutStream> sOutStreams = new ConcurrentHashMap<>();
+  private static Map<Integer, FileInStream> sInStreams = new ConcurrentHashMap<>();
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: we need a way to close and evict stale streams.

File: core/client/src/main/java/alluxio/client/file/options/CreateDirectoryOptions.java [hunk 10]
```diff
@@ -60,8 +58,8 @@ public Mode getMode() {
   /**
    * @return the under storage type
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/CreateDirectoryOptions.java]: done

File: core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java [hunk 11]
```diff
@@ -123,7 +123,6 @@ public boolean hasOwner() {
    * @return the owner
    */
   public String getOwner() {
-    Preconditions.checkState(hasOwner(), PreconditionMessage.MUST_SET_OWNER);
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java]: no

Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java]: and without the serialization of SetAttributeOptions that do not have these options set fails

File: core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java [hunk 12]
```diff
@@ -75,15 +77,25 @@ public long getBlockSizeBytes() {
...
+   */
+  @JsonProperty("locationPolicy")
+  public String getLocationPolicyClass() {
+    return mLocationPolicy.getClass().getCanonicalName();
+  }
+
+  /**
    * @return the Alluxio storage type
    */
+  @JsonIgnore
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java]: done

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java [hunk 13]
```diff
@@ -59,24 +58,34 @@
 @NotThreadSafe
 @Path(FileSystemClientRestServiceHandler.SERVICE_PREFIX)
 @Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
 public final class FileSystemClientRestServiceHandler {
-  public static final String SERVICE_PREFIX = "file";
-  public static final String SERVICE_NAME = "service_name";
-  public static final String SERVICE_VERSION = "service_version";
-  public static final String CREATE_DIRECTORY = "create_directory";
+  public static final String SERVICE_PREFIX = "alluxio";
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: not really, but if we don't then I would need to split up the implementation of the API into two classes, which I am happy to do

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 3 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: any issues with the URLs will be caught by tests which use the suffixes and prefixes and I would rather not duplicated the string constants

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 5 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: nope because the input argument is final as we are using a nested class

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 7 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: fixed everywhere (I use to keep one argument per line when we had a lot of them)

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 4 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: much better, now it is thread-safe

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 9 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: will do

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 6 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: will do

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 3 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: sg, I'm ok with either approach

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 5 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: ah, right

File: tests/src/test/java/alluxio/rest/TestCase.java (hunk 8 above)
Comment (aaudiber) [inline tests/src/test/java/alluxio/rest/TestCase.java]: sg, add a comment?

File: core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java (hunk 13 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: sgtm

Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemClientRestServiceHandler.java]: I will make the change, but it will require some refactoring

File: core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java [hunk 14]
```diff
@@ -75,15 +77,25 @@ public long getBlockSizeBytes() {
   }
 
   /**
-   * @return the location policy for writes to Alluxio storage
+   * @return the location policy used when storing data to Alluxio
    */
+  @JsonIgnore
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java]: discussed offline

File: core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java [hunk 15]
```diff
@@ -75,15 +77,25 @@ public long getBlockSizeBytes() {
...
   public FileWriteLocationPolicy getLocationPolicy() {
     return mLocationPolicy;
   }
 
   /**
+   * @return the location policy class used when storing data to Alluxio
+   */
+  @JsonProperty("locationPolicy")
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/CreateFileOptions.java]: done

File: core/common/src/main/java/alluxio/Constants.java (hunk 2 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Constants.java]: v1/api implies that api is a subset of v1, while it is the other way around

File: tests/src/test/java/alluxio/rest/TestCase.java (hunk 8 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/rest/TestCase.java]: done

File: core/common/src/main/java/alluxio/Constants.java (hunk 2 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/Constants.java]: got it, since the API is changing from 1.3 -> 1.4, we do not need to worry about backward compatibility?

Comment (jsimsa) [inline core/common/src/main/java/alluxio/Constants.java]: it will break backward compatibility of internal REST API 1.3 clients with 1.4, but that is fine with me

**Comment by jsimsa** [comment]
@calvinjia @aaudiber this PR is now ready for another round of review

**Comment by jsimsa** [comment]
Jenkins, test this please

File: core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java (hunk 11 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java]: we should remove the unused `PreconditionMessage` constants

File: core/client/src/test/java/alluxio/client/file/options/CreateFileOptionsTest.java [hunk 16]
```diff
@@ -82,8 +80,6 @@ public void fields() {
     Assert.assertEquals(recursive, options.isRecursive());
     Assert.assertEquals(ttl, options.getTtl());
     Assert.assertEquals(TtlAction.FREE, options.getTtlAction());
-    Assert.assertEquals(writeType.getAlluxioStorageType(), options.getAlluxioStorageType());
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/options/CreateFileOptionsTest.java]: test writeType

File: core/client/src/test/java/alluxio/client/file/options/CreateFileOptionsTest.java [hunk 17]
```diff
@@ -48,8 +48,6 @@ public void defaults() {
     CreateFileOptions options = CreateFileOptions.defaults();
     Assert.assertTrue(options.isRecursive());
     Assert.assertEquals(mDefaultBlockSizeBytes, options.getBlockSizeBytes());
-    Assert.assertEquals(mDefaultWriteType.getAlluxioStorageType(), options.getAlluxioStorageType());
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/options/CreateFileOptionsTest.java]: test writeType

File: core/common/src/main/java/alluxio/Constants.java (hunk 2 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/Constants.java]: are we ok with old versions of manager not being able to talk to new clusters, and vice-versa?

File: core/server/src/main/java/alluxio/proxy/PathsRestServiceHandler.java [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+   *
+   * @param context context for the servlet
+   */
+  public PathsRestServiceHandler(@Context ServletContext context) {
+    mFileSystem =
+        (FileSystem) context.getAttribute(ProxyWebServer.FILE_SYSTEM_SERVLET_RESOURCE_KEY);
+  }
+
+  /**
+   * @summary creates a directory
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/PathsRestServiceHandler.java]: is there a place where we can write a full description that will go into the miredot docs?

File: core/server/src/main/java/alluxio/proxy/PathsRestServiceHandler.java [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+/**
+ * This class is a REST handler for file system client requests.
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/PathsRestServiceHandler.java]: update javadoc now that it's only for path resources

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Stream cache.
+ */
+public final class StreamCache {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final RemovalListener<Integer, Closeable> CLOSER =
+      new RemovalListener<Integer, Closeable>() {
+        public void onRemoval(@Nonnull RemovalNotification<Integer, Closeable> removal) {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: do we need `@Nonnull`?

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+          try {
+            Closeable stream = removal.getValue();
+            if (stream != null) {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: is this possible?

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+      CacheBuilder.newBuilder().expireAfterAccess(1, TimeUnit.HOURS).removalListener(CLOSER)
+          .build();
+  private static Cache<Integer, FileOutStream> sOutStreamCache =
+      CacheBuilder.newBuilder().expireAfterAccess(1, TimeUnit.HOURS).removalListener(CLOSER)
+          .build();
+  private static AtomicInteger sCounter = new AtomicInteger();
+
+  /**
+   * @param key the key to get the stream for
+   * @return the stream
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: , or null if the cache contains no stream with the specified key

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+   */
+  public static FileInStream getInStream(Integer key) {
+    return sInStreamCache.getIfPresent(key);
+  }
+
+  /**
+   * @param key the key to get the stream for
+   * @return the stream
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: ditto

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+   */
+  public static Integer put(FileOutStream os) {
+    int id = sCounter.incrementAndGet();
+    sOutStreamCache.put(id, os);
+    return id;
+  }
+
+  /**
+   * @param key the key for the stream to invalidate
+   * @return the invalidated stream
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: , or null ...

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    FileOutStream os = sOutStreamCache.getIfPresent(key);
+    if (os != null) {
+      sOutStreamCache.invalidate(key);
+      return os;
+    }
+    return null;
+  }
+
+  private StreamCache() {} // prevent instantiation
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: Can we make everything instanced instead of static, and inject the stream cache through the servlet context like we do with the FileSystem?

File: core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+/**
+ * This class is a REST handler for file system client requests.
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java]: update

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Stream cache.
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: more description, e.g. expiration behavior, ownership of stream closing, ID assignment.

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+ */
+public final class StreamCache {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: need a StreamCacheTest, we might want need to make the timeout configurable

Comment (calvinjia) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: Should this class be ThreadSafe?

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+    int id = sCounter.incrementAndGet();
+    sOutStreamCache.put(id, os);
+    return id;
+  }
+
+  /**
+   * @param key the key for the stream to invalidate
+   * @return the invalidated stream
+   */
+  public static Closeable invalidate(Integer key) {
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: Are we leveraging the assumption only 1 thread should have access to the key?

File: core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+   * @summary writes to a stream
+   * @param id the stream id
+   * @param is the input stream
+   * @return the response object
+   */
+  @POST
+  @Path(ID_PARAM + WRITE)
+  @ReturnType("java.lang.Void")
+  @Consumes(MediaType.APPLICATION_OCTET_STREAM)
+  public Response write(@PathParam("id") final Integer id, final InputStream is) {
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java]: Should this return number of bytes successfully written?

File: core/client/src/test/java/alluxio/client/file/options/CreateFileOptionsTest.java (hunk 16 above)
Comment (jsimsa) [inline core/client/src/test/java/alluxio/client/file/options/CreateFileOptionsTest.java]: done

File: core/common/src/main/java/alluxio/Constants.java (hunk 2 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Constants.java]: I think it is fine if clients of the internal v1.3 REST API are not compatible with v1.4 REST API

but newer clients should be still able to talk to v1.3 servers by checking for the version first and the figuring out which API to us

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 27 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 28 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: added configuration setting

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 21 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: ditto

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 22 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 24 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 25 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java (hunk 26 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 28 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: added a test

File: core/server/src/main/java/alluxio/proxy/StreamCache.java [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.Nonnull;
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: remove

File: core/server/src/main/java/alluxio/proxy/PathsRestServiceHandler.java [hunk 32]
```diff
@@ -0,0 +1,375 @@
...
+ */
+@NotThreadSafe
+@Path(PathsRestServiceHandler.SERVICE_PREFIX)
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+public final class PathsRestServiceHandler {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/PathsRestServiceHandler.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 28 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: done

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 29 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: are you asking if we assume that only one thread can be operating over a stream at the time?

File: core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java (hunk 30 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamsRestServiceHandler.java]: done

**Comment by jsimsa** [comment]
@calvinjia @aaudiber addressed your comments, PTAL

File: core/server/src/main/java/alluxio/StreamCache.java [hunk 33]
```diff
@@ -0,0 +1,122 @@
...
+      };
+
+  private AtomicInteger mCounter;
+  private Cache<Integer, FileInStream> mInStreamCache;
+  private Cache<Integer, FileOutStream> mOutStreamCache;
+
+  /**
+   * Creates a new instance of {@link StreamCache}.
+   */
+  public StreamCache() {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/StreamCache.java]: can we take the timeout as a parameter and have the caller look it up in the configuration?

File: core/server/src/main/java/alluxio/proxy/FileSystemRestServiceHandler.java [hunk 34]
```diff
@@ -49,13 +50,13 @@
...
- * This class is a REST handler for file system client requests.
+ * This class is a REST handler for path resources.
  */
 @NotThreadSafe
-@Path(PathsRestServiceHandler.SERVICE_PREFIX)
+@Path(FileSystemRestServiceHandler.SERVICE_PREFIX)
 @Produces(MediaType.APPLICATION_JSON)
 @Consumes(MediaType.APPLICATION_JSON)
-public final class PathsRestServiceHandler {
+public final class FileSystemRestServiceHandler {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/FileSystemRestServiceHandler.java]: I liked `PathsRestServiceHandler` better for consistency with StreamsRestServiceHandler, up to you

File: core/server/src/main/java/alluxio/proxy/StreamRestServiceHandler.java [hunk 35]
```diff
@@ -33,13 +35,13 @@
...
- * This class is a REST handler for file system client requests.
+ * This class is a REST handler for stream resources.
  */
 @NotThreadSafe
-@Path(StreamsRestServiceHandler.SERVICE_PREFIX)
+@Path(StreamRestServiceHandler.SERVICE_PREFIX)
 @Produces(MediaType.APPLICATION_JSON)
 @Consumes(MediaType.APPLICATION_JSON)
-public final class StreamsRestServiceHandler {
+public final class StreamRestServiceHandler {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/proxy/StreamRestServiceHandler.java]: I liked streams better since the prefix is streams. Generally it seems nice to have the REST service handler for resource X be called XRestServiceHandler

File: core/server/src/test/java/alluxio/StreamCacheTest.java [hunk 36]
```diff
@@ -0,0 +1,54 @@
...
+
+public class StreamCacheTest {
+  @Test
+  public void operations() throws Exception {
+    StreamCache streamCache = new StreamCache();
+    FileInStream is = Mockito.mock(FileInStream.class);
+    FileOutStream os = Mockito.mock(FileOutStream.class);
+    Integer isId = streamCache.put(is);
+    Integer osId = streamCache.put(os);
+    Assert.assertNotNull(streamCache.getInStream(isId));
```
Comment (aaudiber) [inline core/server/src/test/java/alluxio/StreamCacheTest.java]: assertSame(is, streamCache.getInStream(isId))

File: core/server/src/test/java/alluxio/StreamCacheTest.java [hunk 37]
```diff
@@ -0,0 +1,54 @@
...
+    Assert.assertNull(streamCache.getInStream(osId));
+    Assert.assertNull(streamCache.getOutStream(isId));
+    Assert.assertNotNull(streamCache.getOutStream(osId));
+    Assert.assertNotNull(streamCache.invalidate(isId));
+    Assert.assertNotNull(streamCache.invalidate(osId));
+    Assert.assertNull(streamCache.invalidate(isId));
+    Assert.assertNull(streamCache.invalidate(osId));
+    Mockito.verify(is, Mockito.times(1)).close();
```
Comment (aaudiber) [inline core/server/src/test/java/alluxio/StreamCacheTest.java]: Mockito.times(1) is implicit

File: core/server/src/main/java/alluxio/proxy/FileSystemRestServiceHandler.java (hunk 34 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/FileSystemRestServiceHandler.java]: @calvinjia changing back to PathsRestServiceHandler, if you feel strongly about it, battle it out with @aaudiber

File: core/server/src/main/java/alluxio/proxy/StreamRestServiceHandler.java (hunk 35 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/proxy/StreamRestServiceHandler.java]: done

File: core/server/src/test/java/alluxio/StreamCacheTest.java (hunk 36 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/StreamCacheTest.java]: done

File: core/server/src/test/java/alluxio/StreamCacheTest.java (hunk 37 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/StreamCacheTest.java]: done

**Comment by jsimsa** [comment]
PTAL

File: core/server/src/main/java/alluxio/proxy/StreamCache.java (hunk 29 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/proxy/StreamCache.java]: Yes, the stream should not be accessed by multiple threads (clients) at the same time right?

**Comment by calvinjia** [comment]
LGTM

File: core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java (hunk 11 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/SetAttributeOptions.java]: the SetAttribute options class is the only class when null makes sense as the default value, for other options we actually want to provide defaults ... so I don't think this change needs to be propagated
